kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/KafkaService.java

```diff
@@ -674,4 +679,154 @@ private InternalClusterMetrics calculateClusterMetrics(
     }
   }
 
+  private Mono<InternalTopic> changeReplicationFactor(
+      AdminClient adminClient,
+      String topicName,
+      Map<TopicPartition, Optional<NewPartitionReassignment>> reassignments
+  ) {
+    return ClusterUtil.toMono(adminClient
+        .alterPartitionReassignments(reassignments).all(), topicName)
+        .flatMap(topic -> getTopicsData(adminClient, Collections.singleton(topic)).next());
+  }
+
+  public Mono<InternalTopic> changeReplicationFactor(
+      KafkaCluster cluster,
+      String topicName,
+      ReplicationFactorChange replicationFactorChange) {
+    return getOrCreateAdminClient(cluster)
+        .flatMap(ac -> {
+          Integer actual = cluster.getTopics().get(topicName).getReplicationFactor();
+          Integer requested = replicationFactorChange.getTotalReplicationFactor();
+          Integer brokersCount = cluster.getMetrics().getBrokerCount();
+
+          if (requested.equals(actual)) {
+            return Mono.error(
+                new ValidationException(
+                    String.format("Topic already has replicationFactor %s.", actual)));
+          }
+          if (requested > brokersCount) {
+            return Mono.error(
+                new ValidationException(
+                    String.format("Requested replication factor %s more than brokers count %s.",
+                        requested, brokersCount)));
+          }
+          return changeReplicationFactor(ac.getAdminClient(), topicName,
+              getPartitionsReassignments(cluster, topicName,
+                  replicationFactorChange));
+        });
+  }
+
+  private Map<TopicPartition, Optional<NewPartitionReassignment>> getPartitionsReassignments(
+      KafkaCluster cluster,
+      String topicName,
+      ReplicationFactorChange replicationFactorChange) {
+    Map<TopicPartition, Optional<NewPartitionReassignment>> reassignments = new HashMap<>();
+
+    // Difference between requested and actual Replication factor
+    Integer replicasCountDiff = replicationFactorChange.getTotalReplicationFactor()
+        - cluster.getTopics().get(topicName).getReplicationFactor();
+
+    Map<Integer, List<Integer>> currentAssignment = new HashMap<>();
+
+    // If we should to increase Replication factor
+    if (replicasCountDiff > 0) {
+
+      // Map for store brokers (Entry<brokerId, topicReplicasCount>)
+      Map<Integer, Integer> brokers = new LinkedHashMap<>();
+
+      // Add available brokers to map
+      cluster.getMetrics().getInternalBrokerDiskUsage().keySet()
+          .forEach(id -> brokers.put(id, 0));
+
+      // Fill currentAssignment and brokers
+      cluster.getTopics().get(topicName).getPartitions().values()
```

Comment:
please simplify this block, it's a bit hard to understand what is happening here

Reply:
Please check again, now should be better